backend/cellguide/pipeline/computational_marker_genes/utils.py

```diff
@@ -193,3 +91,69 @@ def query_gene_info_for_gene_description(gene_id: str) -> str:
         return data["name"]
     else:
         return gene_id
+
+
+@njit(parallel=True)
+def bootstrap_rows_percentiles(
+    X: np.ndarray, num_replicates: int = 1000, num_samples: int = 100, percentile: float = 5
+):
+    """
+    This function bootstraps rows of a given matrix X.
+
+    Arguments
+    ---------
+    X : np.ndarray
+        The input matrix to bootstrap.
+    num_replicates : int, optional
+        The number of bootstrap replicates to generate, by default 1000.
+    num_samples : int, optional
+        The number of samples to draw in each bootstrap replicate, by default 100.
+    percentile : float, optional
+        The percentile of the bootstrapped samples for each replicate, by default 15.
+
+    Returns
+    -------
+    bootstrap_percentile : np.ndarray
+        The percentile of the bootstrapped samples for each replicate.
+    """
+
+    bootstrap_percentile = np.zeros((num_replicates, X.shape[1]), dtype="float")
+
+    indices = np.random.randint(0, X.shape[0], size=(num_replicates, num_samples))
+    # for each replicate
+    for n_i in prange(num_replicates):
+        bootstrap_percentile[n_i] = sort_matrix_columns(X[indices[n_i]], percentile, num_samples)
+
+    return bootstrap_percentile
+
+
+@njit
+def sort_matrix_columns(matrix, percentile, num_samples):
```

Comment:
This function is used to approximate nanpercentile. In fact, it is virtually the same as how percentiles are typically found except without linear interpolation. This approximation is much faster than nanpercentile.

When comparing this implementation to nanpercentile, the maximum numerical deviation out of the millions of resulting values is 8%. Typically, the deviation is much smaller at <1%.

Reply:
Consider pushing your improvements upstream to numpy, rather than re-implementing our own thing here.